### DISCUSSION TITLE
A way to prevent expand_auto() for some time while typing

### DISCUSSION THREAD

**Discussion by 5721309** [L3MON4D3/LuaSnip#761] [Ideas]
Maybe there is a simple DIY solution like to assign an empty function to `expand_auto`, or to reconfigure luasnip on the fly (is it possible?).
But I feel you could provide us a function which receives boolean and accordingly enables or disables `expand_auto`. What do you think?

The bigger problem is that when you add autosnippets like `if` or `while` you believe that they shouldn't expand in some environments like `@string` or `@comment`. Using treesitter api seems hard for that - I haven't found any tutorial for that (have you?).

**Comment by L3MON4D3** [comment]
Use `condition`!
Dont have a link on hand, but it's in the DOC in the `Snippet`-section

**Reply by 5721309** [reply]
Thank you, yes It seems I can use it. Here is the [link to the DOC](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snippets). We can pass conditions via `opts` parameter of `s()`.

**Comment by 5721309** [comment]
Write this in e.g. lua/snippets_common.lua
```lua
local M = {}

local expand_auto_allowed = true

vim.keymap.set({'i', 'n'}, '<A-a>', function() expand_auto_allowed = not expand_auto_allowed end)

local function condition_for_expand_auto()
	return expand_auto_allowed
end

M.expand_auto_condition = require("luasnip.extras.conditions").make_condition(condition_for_expand_auto)

return M
```
And then in your snippet files:
```lua
local eac = require('snippets_common').expand_auto_condition
-- ...
s('if ', fmt('if {} then\n\t{}\nend', { i(1), i(0) }), {condition = eac})
```

**Reply by L3MON4D3** [reply]
Nice :+1:
You might want to take a look at [`extend_decorator`](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#extend_decorator), with that you don't have to always append the `condition = eac`-part

**Reply by 5721309** [reply]
Ok, using `extend_decorator` looks easier, but I get an error when trying to apply it to `s()`:
```lua
local s_eac = ls.extend_decorator.apply(s, {condition = eac})
-- ERROR : key found in more than one map: condition
```
Can you help me understand why that happens?

**Reply by 5721309** [reply]
Nevermind, I think I understood. `s()` was registered in the way, that second argument to `apply()` is overwriting first parameter of `s()` and the next one overwrites the `opts` parameter.
```lua
local s_eac = ls.extend_decorator.apply(s, nil, {condition = eac})
-- ...
s_eac('if ', fmt('if {} then\n\t{}\nend', { i(1), i(0) }))
```

**Reply by L3MON4D3** [reply]
Yeeah exactly👍👍
